src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -2792,12 +2828,19 @@ bool QgsOgrProvider::changeAttributeValues( const QgsChangedAttributesMap &attr_
             if ( OGR_Fld_GetSubType( fd ) == OFSTJSON )
             {
               stringValue = QString::fromStdString( QgsJsonUtils::jsonFromVariant( it2.value() ).dump() );
+              // setting JSON fields does not work with UpdateFeature as of GDAL 3.8.5
```

Comment:
@rouault is this a known bug?

Reply:
with which driver?

Reply:
Both issues were with JSON, from the QGIS tests in test_provider_ogr.py

Eg https://github.com/qgis/QGIS/blob/master/tests/src/python/test_provider_ogr.py#L1021

Reply:
Actually trying, I realize that UpdateFeature() is totally broken for the GeoJSON driver (the reason is that OGRGeoJSONLayer subclasses OGRMemLayer, but must not forget to override methods to add special processing in case the driver loads features on demand...), not just for those data types. Fix queued in https://github.com/OSGeo/gdal/pull/10197. Actually same issue with ODS and XLSX drivers. Will be fixed in 3.9.1

By the way, when using the UpdateFeature() implementation you could save doing the GetFeature() call at src/core/providers/ogr/qgsogrprovider.cpp#L2646 (but you should do a OGR_F_SetFID() call then). And trying to use UpdateFeature() is going to be only beneficial if the layer advertizes the OLCUpdateFeature capability (if it does not, the default implementation at https://github.com/OSGeo/gdal/blob/6bd9d23a3d3f9bbba5a7b5dfe062cd58f2dee6e7/ogr/ogrsf_frmts/generic/ogrlayer.cpp#L901  is used). So perhaps you should restrict using UpdateFeature() to layers that declare OLCUpdateFeature, *and* for GDAL <= 3.9.0, exclude GeoJSON, ODS and XLSX drivers (they will advertize that capability despite their faulty (non-)implementation since OGRMemLayer does)

Reply:
Ok, thanks! How's this look now?

Reply:
Doh this was still broken in GDAL 3.9.1 and 3.9.2 for GeoJSON... Fixes in https://github.com/OSGeo/gdal/pull/10794 and https://github.com/qgis/QGIS/pull/58740